### DISCUSSION TITLE
How does SumatraPDF store passwords?

### DISCUSSION THREAD

**Discussion by vishalnandagopal** [sumatrapdfreader/sumatrapdf#4354] [General]
When I click click on "save password" when typing the password for a file, a parameter `DecryptionKey` appears in `SumatraPDF-settings.txt`.

What is the encryption process? Is it just something like base64?

**Comment by kjk** [comment]
Possibly hex-encoded but really I forgot the details. You can always look at the code.

**Reply by vishalnandagopal** [reply]
https://github.com/sumatrapdfreader/sumatrapdf/blob/96fca37426eccacbad5b59c38cb592d0dd3a4a39/do/settings_def.go#L168-L172

"Hex encoded MD5 fingerprint of file content (32 chars) followed by "+"crypt key (64 chars) - only applies for PDF documents"

So the first part is clear and I manually tested it, do you know what the crypt key is?

**Reply by kjk** [reply]
I'm pretty sure it's password

**Comment by GitHubRulesOK** [comment]
@vishalnandagopal 
It is not as simple as "base64" since coded length is no indication of number of characters
one letter password is
36280035e9f422e34c31b7cc455ac71026027a544f3be5fbe55781f65943bd9200000000000000000000000000000000
and eight letters is
c07adc198d89bee6bf53ea292f7c3217fe91f15bda233bc15dce25bde5e4abbb00000000000000000000000000000000

4 digits is easier
dc4eabcc07eef6f69f0d33d0c29457635139b4213a000000000000000000000000000000000000000000000000000000
but it is not the PDF file internal hash key which is longer 
E72D0940E6DD12AF935A08BEA7B46B2CDA6F53F3173A9F403602F0ABD34DCEE8

so it may be a simpler cypher reflecting user input.
However comments in code says
>
// Store a password securely (if data encryption is provided by OS)   
// or obfuscated to make search for password in memory dump less trivial.  

and in part may be from MuPDF code base too as that says "padding"

**Reply by vishalnandagopal** [reply]
Okay thanks, will take a look at the code and see if I can figure it out

**Comment by CarolejaneCosmos** [comment]
When I choose to remember the PDF password option, why does it not work? I still need to paste the password each time I open the PDF file.

**Reply by GitHubRulesOK** [reply]
@CarolejaneCosmos 
it should be working using the settings file so for example open this file
https://sample-files.com/downloads/documents/pdf/protected.pdf
**ENSURE** remember is checked and enter the password **samplefiles**
![image](https://github.com/user-attachments/assets/f135e906-3774-4f9b-aa3d-8a6a3989fef0)
close the file and re-open it will not ask for the password HOWEVER the key is stored in settings
```
FileStates [
	[
		FilePath = C:\Users\WDAGUtilityAccount\AppData\Local\Microsoft\Windows\INetCache\IE\QDGP8XLK\protected[1].pdf
		Favorites [
		]
		IsPinned = false
		IsMissing = false
		OpenCount = 1
		DecryptionKey = 5586ea9e1a1c2f03e111044d1264cb6d7f89852d685e8e363ff01cd75f81df7d00000000000000000000000000000000

...
etc.
```
BUT that means the filename must be the same when reopened and importantly remember file settings must also be kept active between sessions
![image](https://github.com/user-attachments/assets/607e29c3-404b-4e06-b7b3-7aae0e1ebd48)

If inactive the option should be greyed out
![image](https://github.com/user-attachments/assets/a5952b52-395a-430c-81c8-d560e8de90d5)

The alternative is keep the password as "plain text" with this advanced option. and then the password is tried against any password based file(s)

![image](https://github.com/user-attachments/assets/db1316e6-deb1-4d0f-8d83-33a115a31e47)
